test/py/test_migration.py

```diff
@@ -28,139 +29,504 @@
 
-    ret = vfu_setup_region(ctx, index=VFU_PCI_DEV_MIGR_REGION_IDX,
-                           size=2 << PAGE_SHIFT, flags=VFU_REGION_FLAG_RW)
-    assert ret == 0
+    cbs = vfu_migration_callbacks_t()
+    cbs.version = 1
+    cbs.transition = migr_trans_cb
+    cbs.read_data = migr_read_data_cb
+    cbs.write_data = migr_write_data_cb
 
-    ret = vfu_setup_device_migration_callbacks(ctx)
+    ret = vfu_setup_device_migration_callbacks(ctx, cbs)
+    assert ret < 0
+
+    cbs.version = VFU_MIGR_CALLBACKS_VERS
+    ret = vfu_setup_device_migration_callbacks(ctx, cbs)
     assert ret == 0
 
     vfu_setup_device_quiesce_cb(ctx)
 
     ret = vfu_realize_ctx(ctx)
     assert ret == 0
 
-    sock = connect_client(ctx)
+    sock = connect_client(ctx, 4)
 
 
-def teardown_function(function):
-    global ctx
-    vfu_destroy_ctx(ctx)
+def get_server_shortest_path(a, b, expectA=0, expectB=0):
+    global path
 
+    if current_state == VFIO_USER_DEVICE_STATE_STOP_COPY and \
+            a == VFIO_USER_DEVICE_STATE_PRE_COPY:
+        # The transition STOP_COPY -> PRE_COPY is explicitly blocked so we
+        # advance one state to get around this in order to set up the test.
+        transition_to_state(VFIO_USER_DEVICE_STATE_STOP)
 
-@patch('libvfio_user.quiesce_cb')
-@patch('libvfio_user.migr_trans_cb')
-def test_migration_bad_access(mock_trans, mock_quiesce):
-    """
-    Tests that attempting to access the migration state register in an
-    non-aligned manner fails.
+    transition_to_state(a, expect=expectA)
 
-    This test is important because we tell whether we need to quiesce by
-    checking for a register-sized access, otherwise we'll change migration
-    state without having quiesced.
-    """
-    global ctx, sock
+    if expectA != 0:
+        return None
+
+    path = []
 
-    data = VFIO_DEVICE_STATE_V1_SAVING.to_bytes(c.sizeof(c.c_int), 'little')
-    write_region(ctx, sock, VFU_PCI_DEV_MIGR_REGION_IDX, offset=0,
-                 count=len(data)-1, data=data, expect=errno.EINVAL)
+    transition_to_state(b, expect=expectB)
 
-    mock_trans.assert_not_called()
+    return path.copy()
 
 
-@patch('libvfio_user.quiesce_cb')
-@patch('libvfio_user.migr_trans_cb', return_value=0)
-def test_migration_trans_sync(mock_trans, mock_quiesce):
+def test_migration_shortest_state_transition_paths():
     """
-    Tests transitioning to the saving state.
+    The spec dictates that complex state transitions are to be implemented as
+    combinations of the defined direct transitions, with the path selected
+    according to the following rules:
+
+    - Select the shortest path.
+    - The path cannot have saving group states as interior arcs, only start/end
+      states.
+
+    This test implements a breadth-first search to ensure that the paths taken
+    by the implementation correctly follow these rules.
     """
 
-    global ctx, sock
+    # states (vertices)
+    V = {VFIO_USER_DEVICE_STATE_ERROR, VFIO_USER_DEVICE_STATE_STOP,
```

Comment:
Wouldn't that involve referring to all the states as `State.VFIO_USER_DEVICE_STATE_*` everywhere or something like that?

Reply:
Actually we can use `E.keys()` like you suggested below.

Reply:
> Wouldn't that involve referring to all the states as State.VFIO_USER_DEVICE_STATE_* everywhere or something like that?

It could be `VFIO_USER_DEVICE_STATE_.STOP`, `VFIO_USER_DEVICE_STATE_.PRE_COPY`, etc. Regardless of using `E.keys()`, it nice to have.